plonky2/src/gates/coset_interpolation.rs

```diff
@@ -0,0 +1,783 @@
+/// weights and z be the evaluation point. Define the sequences
+///
+/// p[0] = 1
+/// p[i] = p[i - 1] * (z - x[i - 1])
+/// e[0] = 0,
+/// e[i] = e[i - 1] * (z - x[i - 1]) + w[i - 1] * v[i - 1] * p[i - 1]
+///
+/// Then e[N] is the final interpolated value. The non-routed wires hold every (d - 1)'th
+/// intermediate value of p and e, starting at p[d] and e[d], where d is the gate degree.
+#[derive(Copy, Clone, Debug)]
+pub struct CosetInterpolationGate<F: RichField + Extendable<D>, const D: usize> {
+    pub subgroup_bits: usize,
+    pub degree: usize,
+    _phantom: PhantomData<F>,
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> InterpolationGate<F, D>
+    for CosetInterpolationGate<F, D>
+{
+    fn new(subgroup_bits: usize) -> Self {
+        Self::with_max_degree(subgroup_bits, 1 << subgroup_bits)
+    }
+
+    fn num_points(&self) -> usize {
+        1 << self.subgroup_bits
+    }
+
+    /// Wire indices of the interpolant's `i`th coefficient.
+    fn wires_coeff(&self, _i: usize) -> Range<usize> {
+        panic!("No coefficient wires");
+    }
+
+    fn end_coeffs(&self) -> usize {
+        self.start_coeffs()
+    }
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> CosetInterpolationGate<F, D> {
+    pub(crate) fn with_max_degree(subgroup_bits: usize, max_degree: usize) -> Self {
+        assert!(max_degree > 1, "need at least quadratic constraints");
+
+        let n_points = 1 << subgroup_bits;
+
+        // Number of intermediate values required to compute interpolation with degree bound
+        let n_intermediates = (n_points - 2) / (max_degree - 1);
+
+        // Find minimum degree such that (n_points - 2) / (degree - 1) < n_intermediates + 1
+        // Minimizing the degree this way allows the gate to be in a larger selector group
+        let degree = (n_points - 2) / (n_intermediates + 1) + 2;
+
+        Self {
+            subgroup_bits,
+            degree,
+            _phantom: PhantomData,
+        }
+    }
+
+    fn num_intermediates(&self) -> usize {
+        (self.num_points() - 2) / (self.degree() - 1)
+    }
+
+    /// The wires corresponding to the i'th intermediate evaluation.
+    fn wires_intermediate_eval(&self, i: usize) -> Range<usize> {
+        debug_assert!(i < self.num_intermediates());
+        let start = self.end_coeffs() + D * i;
+        start..start + D
+    }
+
+    /// The wires corresponding to the i'th intermediate product.
+    fn wires_intermediate_prod(&self, i: usize) -> Range<usize> {
+        debug_assert!(i < self.num_intermediates());
+        let start = self.end_coeffs() + D * (self.num_intermediates() + i);
+        start..start + D
+    }
+
+    fn barycentric_weights(&self) -> Vec<F> {
+        barycentric_weights(
+            &F::two_adic_subgroup(self.subgroup_bits)
+                .into_iter()
+                .map(|x| (x, F::ZERO))
+                .collect::<Vec<_>>(),
+        )
+    }
```

Comment:
nit: Could we have `barycentric_weights: Vec<F>` be a field of `CosetInterpolationGate` instead? To be sure they're computed only once.

Reply:
The `InterpolationGate` trait inherits from Copy, so it can't hold a Vec. I'm not sure if that could be removed or how much refactoring it would require.

Reply:
Oh right. Then it's fine to leave it like this for now, I can work on refactoring the `InterpolationGate` trait in another PR.